addons/mail/static/tests/thread_window/moderation_thread_window_tests.js

```diff
@@ -0,0 +1,137 @@
+                channel_slots: {
+                    channel_channel: [{
+                        id: 1,
+                        channel_type: "channel",
+                        name: "general",
+                    }],
+                },
+            },
+        };
+        this.services = mailTestUtils.getMailServices();
+        this.ORIGINAL_THREAD_WINDOW_APPENDTO = this.services.mail_service.prototype.THREAD_WINDOW_APPENDTO;
+
+        this.createParent = function (params) {
+            var widget = new Widget();
+
+            // in non-debug mode, append thread windows in qunit-fixture
+            // note that it does not hide thread window because it uses fixed
+            // position, and qunit-fixture uses absolute...
+            if (params.debug) {
+                self.services.mail_service.prototype.THREAD_WINDOW_APPENDTO = 'body';
+            } else {
+                self.services.mail_service.prototype.THREAD_WINDOW_APPENDTO = '#qunit-fixture';
+            }
+
+            testUtils.addMockEnvironment(widget, params);
+            return widget;
+        };
+    },
+    afterEach: function () {
+        // reset thread window append to body
+        this.services.mail_service.prototype.THREAD_WINDOW_APPENDTO = 'body';
+    },
+});
+
+QUnit.test('moderator: moderated channel with pending moderation message', function (assert) {
+    assert.expect(5);
+
+    this.data.initMessaging = {
+        channel_slots: {
+            channel_channel: [{
+                id: 1,
+                channel_type: "channel",
+                name: "general",
+                moderation: true,
+            }],
+        },
+        is_moderator: true,
+        moderation_counter: 1,
+        moderation_channel_ids: [1],
+    };
+    this.data['mail.message'].records = [{
+        author_id: [2, "Someone"],
+        body: "<p>test</p>",
+        id: 100,
+        model: 'mail.channel',
+        moderation_status: 'pending_moderation',
+        need_moderation: true,
+        res_id: 1,
+    }];
+
+    var parent = this.createParent({
+        data: this.data,
+        services: this.services,
+    });
+
+    // detach channel 1, so that it opens corresponding thread window.
+    parent.call('mail_service', 'getChannel', 1).detach();
+
+    assert.strictEqual($('.o_thread_message ').length, 1,
+        "should display a message in the thread window");
+    assert.strictEqual($('.o_thread_author').text().trim(), "Someone",
+        "should be a message of 'Someone'");
+    assert.strictEqual($('.o_thread_icons').text().trim(), "Pending moderation",
+        "should be pending moderation");
+    assert.strictEqual($('.o_thread_message_moderation').length, 0,
+        "should not display any contextual moderation decisions next to the message");
+    assert.strictEqual($('.moderation_checkbox').length, 0,
+        "should not display any moderation checkbox next to the message");
+
+    parent.destroy();
+});
+
+});
+});
+});
```

Comment:
Is there a test of chat window for an author non-moderator?

Reply:
No test for chat window for author non-moderator, but there is a discuss test for this. I don't think this is important to test that, otherwise more and more tests must be duplicated in chat windows from discuss.